File: src/main/java/ru/mdashlw/rankedwho/command/Command.java
package ru.mdashlw.rankedwho.command;

import net.minecraft.command.ICommand;
import net.minecraft.command.ICommandSender;
import net.minecraft.util.BlockPos;
import net.minecraft.util.ChatComponentText;
import net.minecraft.util.ChatStyle;
import net.minecraft.util.EnumChatFormatting;
import net.minecraftforge.client.ClientCommandHandler;
import ru.mdashlw.rankedwho.RankedWho;

import java.util.Collections;
import java.util.List;

public abstract class Command implements ICommand {
    protected RankedWho mod;

    public Command(RankedWho mod) {
        this.mod = mod;
    }

    public abstract String getName();

    public List<String> getAliases() {
        return Collections.emptyList();
    }

    public boolean isAsync() {
        return false;
    }

    public void register() {
        ClientCommandHandler.instance.registerCommand(this);
    }

    public abstract void execute(ICommandSender sender, String[] args);

    @Override
    public String getCommandName() {
        return getName();
    }

    @Override
    public String getCommandUsage(ICommandSender sender) {
        return "";
    }

    @Override
    public List<String> getCommandAliases() {
        return getAliases();
    }

    @Override
    public void processCommand(ICommandSender sender, String[] args) {
        if (isAsync()) {
            mod.getCommandHandler().submit(() -> {
                run(sender, args);
            });
        } else {
            run(sender, args);
        }
    }

    private void run(ICommandSender sender, String[] args) {
        try {
            execute(sender, args);
        } catch (Throwable exception) {
            exception.printStackTrace();
            sender.addChatMessage(new ChatComponentText("Uncaught exception has been thrown while executing this command, contact the mod owner.")
                    .setChatStyle(new ChatStyle()
                            .setColor(EnumChatFormatting.RED)));
        }
    }

    @Override
    public boolean canCommandSenderUseCommand(ICommandSender sender) {
        return true;
    }

    @Override
    public List<String> addTabCompletionOptions(ICommandSender sender, String[] args, BlockPos pos) {
        return null;
    }

    @Override
    public boolean isUsernameIndex(String[] args, int index) {
        return false;
    }

    @Override
    public int compareTo(ICommand o) {
        return getName().compareTo(o.getCommandName());
    }
}

File: src/main/kotlin/ru/mdashlw/rankedwho/command/Command.kt
package ru.mdashlw.rankedwho.command

import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import net.minecraft.command.ICommand
import net.minecraft.command.ICommandSender
import net.minecraft.util.BlockPos
import net.minecraftforge.client.ClientCommandHandler
import ru.mdashlw.rankedwho.RankedWho

abstract class Command : ICommand {
    abstract val aliases: List<String>
    open val async: Boolean = false

    fun register() {
        ClientCommandHandler.instance.registerCommand(this)
    }

    abstract fun execute(args: List<String>)

    override fun getCommandName(): String = aliases.first()

    override fun getCommandUsage(sender: ICommandSender?): String = ""

    override fun getCommandAliases(): MutableList<String> = aliases.drop(1).toMutableList()

    @Suppress("NAME_SHADOWING")
    override fun processCommand(sender: ICommandSender, args: Array<out String>) {
        val args = args.toList()

        if (async) {
            GlobalScope.launch(RankedWho.singlePool) {
                execute(args)
            }
        } else {
            execute(args)
        }
    }

    override fun canCommandSenderUseCommand(sender: ICommandSender): Boolean = true

    override fun addTabCompletionOptions(
        sender: ICommandSender?,
        args: Array<out String>?,
        pos: BlockPos?
    ): MutableList<String>? = null

    override fun isUsernameIndex(args: Array<out String>?, index: Int): Boolean = false

    override fun compareTo(other: ICommand): Int = commandName.compareTo(other.commandName)
}
